### PR TITLE
LPS-190826 Add value field to grid response

### PR DESCRIPTION
**Pull request by YurenaC**
Motivation
=======
An impedibug was raised when testing a relates story as when the user customizes the option labels available in the grid sections in a web content, those new names are not shown properly. (Please for more info go to the original [Story](https://liferay.atlassian.net/browse/LPS-190826) ) 

Proposed Solution
========
In order to keep showing accurate information when a graphQL query is executed, information like the customized field reference value (label) and the original reference will be added to the response.

Steps to Verify
========

1. Add a web content structure with a Grid field
2. Add row and column options with customized field reference for Grid
3. Add a web content based on new structure
4. Get the id of structuredContent via getSiteStructuredContentsPage
5. Navigate to the GraphQL
6. Get the structuredContent with contentFieldValue

 ```

{
  structuredContent(structuredContentId: {structuredContentId}) {
    contentFields {
			contentFieldValue {
			  data
			  value
			}
    }
  }
}
```
7. Expected result as follows:
GridExpectedResult

COMMITS:
-------------------
LPS-190226 Add value field to grid response

### CODE REVIEW

**Comment by YurenaC** [comment]
ci:test:sf

**Comment by YurenaC** [comment]
ci:test:sf

File: modules/apps/headless/headless-delivery/headless-delivery-api/src/main/java/com/liferay/headless/delivery/dto/v1_0/util/ContentFieldUtil.java [hunk 1]
```diff
@@ -374,6 +374,58 @@ else if (Objects.equals(
...
+				return new ContentFieldValue() {
+					{
+						setData(
+							String.valueOf(
+								JSONFactoryUtil.createJSONObject(
+									localizedSelectedData)));
+						setValue(
+							String.valueOf(
+								JSONFactoryUtil.createJSONObject(
+									selectedValues)));
```
[the remark this answers is not in the document]
Comment (YurenaC) [reply modules/apps/headless/headless-delivery/headless-delivery-api/src/main/java/com/liferay/headless/delivery/dto/v1_0/util/ContentFieldUtil.java:425]: I'm using a JSONObject in order to show the data as expected.
Initially, I also used the Map and the data was showed as key=value and we expected key:value

File: modules/apps/headless/headless-delivery/headless-delivery-api/src/main/java/com/liferay/headless/delivery/dto/v1_0/util/ContentFieldUtil.java [hunk 2]
```diff
@@ -374,6 +374,58 @@ else if (Objects.equals(
...
+						"rows"
+					);
+				DDMFormFieldOptions columnsFieldOptions =
+					(DDMFormFieldOptions)ddmFormField.getProperties(
+					).get(
+						"columns"
+					);
+
+				Map<String, String> selectedData = JSONUtil.toStringMap(
+					JSONFactoryUtil.createJSONObject(valueString));
```
[the remark this answers is not in the document]
Comment (YurenaC) [reply modules/apps/headless/headless-delivery/headless-delivery-api/src/main/java/com/liferay/headless/delivery/dto/v1_0/util/ContentFieldUtil.java:393]: It simplifies the iteration needed a few lines below I can change it and leave it as a JSONObject but in the iteration I'll have to call the toMap() method anyway (selectedData.toMap().forEach(...))
What do you think is the cleanest approach?

File: modules/apps/headless/headless-delivery/headless-delivery-api/src/main/java/com/liferay/headless/delivery/dto/v1_0/util/ContentFieldUtil.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (lfbesada) [reply modules/apps/headless/headless-delivery/headless-delivery-api/src/main/java/com/liferay/headless/delivery/dto/v1_0/util/ContentFieldUtil.java:425]: The correct way to convert a JSONObject in String is using JSONUtil.toString(jsonObject).
In addition, I don't understand the reason for creating a string Map and later converting it to JSONOject, I would prefer to use JSONObject directly.
What do you think?

File: modules/apps/headless/headless-delivery/headless-delivery-api/src/main/java/com/liferay/headless/delivery/dto/v1_0/util/ContentFieldUtil.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (lfbesada) [reply modules/apps/headless/headless-delivery/headless-delivery-api/src/main/java/com/liferay/headless/delivery/dto/v1_0/util/ContentFieldUtil.java:393]: Why do you need the forEach? You can iterate over the JSON using a for loop as we always do everywhere

**Comment by lfbesada** [comment]
ci:test:sf

File: modules/apps/headless/headless-delivery/headless-delivery-api/src/main/java/com/liferay/headless/delivery/dto/v1_0/util/ContentFieldUtil.java [hunk 3]
```diff
@@ -374,6 +374,58 @@ else if (Objects.equals(
 					}
 				};
 			}
+			else if (Objects.equals(
+						DDMFormFieldTypeConstants.GRID,
+						ddmFormField.getType())) {
```
[the remark this answers is not in the document]
Comment (lfbesada) [reply modules/apps/headless/headless-delivery/headless-delivery-api/src/main/java/com/liferay/headless/delivery/dto/v1_0/util/ContentFieldUtil.java:396]: Could you please sort this clause alphabetically?

File: modules/apps/headless/headless-delivery/headless-delivery-api/src/main/java/com/liferay/headless/delivery/dto/v1_0/util/ContentFieldUtil.java [hunk 4]
```diff
@@ -390,6 +391,55 @@ else if (Objects.equals(
...
+				while (iterator.hasNext()) {
+					String key = iterator.next();
+
+					String value = jsonObject.getString(key);
+
+					LocalizedValue optionKeyLabels =
+						rowsFieldOptions.getOptionLabels(key);
+
+					LocalizedValue optionValueLabels =
+						columnsFieldOptions.getOptionLabels(value);
```
Comment (lfbesada) [inline modules/apps/headless/headless-delivery/headless-delivery-api/src/main/java/com/liferay/headless/delivery/dto/v1_0/util/ContentFieldUtil.java:424]: The name of these variables should end with "LocalizedValue"

File: modules/apps/headless/headless-delivery/headless-delivery-api/src/main/java/com/liferay/headless/delivery/dto/v1_0/util/ContentFieldUtil.java [hunk 5]
```diff
@@ -390,6 +391,55 @@ else if (Objects.equals(
...
+			else if (Objects.equals(
+						DDMFormFieldTypeConstants.GRID,
+						ddmFormField.getType())) {
+
+				Map<String, Object> properties = ddmFormField.getProperties();
+
+				DDMFormFieldOptions rowsFieldOptions =
+					(DDMFormFieldOptions)properties.get("rows");
+				DDMFormFieldOptions columnsFieldOptions =
+					(DDMFormFieldOptions)properties.get("columns");
```
Comment (lfbesada) [inline modules/apps/headless/headless-delivery/headless-delivery-api/src/main/java/com/liferay/headless/delivery/dto/v1_0/util/ContentFieldUtil.java:403]: The name of these variables should end with "DDMFormFieldOptions"

File: modules/apps/headless/headless-delivery/headless-delivery-api/src/main/java/com/liferay/headless/delivery/dto/v1_0/util/ContentFieldUtil.java [hunk 6]
```diff
@@ -390,6 +391,55 @@ else if (Objects.equals(
...
+					selectedValuesJSONObject.put(
+						rowsFieldOptions.getOptionReference(key),
+						columnsFieldOptions.getOptionReference(value));
+				}
+
+				return new ContentFieldValue() {
+					{
+						setData(
+							String.valueOf(localizedSelectedDataJSONObject));
+						setValue(String.valueOf(selectedValuesJSONObject));
```
Comment (lfbesada) [inline modules/apps/headless/headless-delivery/headless-delivery-api/src/main/java/com/liferay/headless/delivery/dto/v1_0/util/ContentFieldUtil.java:439]: The correct way to convert a JSONObject in String is using JSONUtil.toString(jsonObject).

**Review by lfbesada — CHANGES_REQUESTED** [review]

**Comment by YurenaC** [comment]
ci:test:sf

**Comment by YurenaC** [comment]
ci:test:relevant

File: modules/apps/headless/headless-delivery/headless-delivery-api/src/main/java/com/liferay/headless/delivery/dto/v1_0/util/ContentFieldUtil.java [hunk 7]
```diff
@@ -220,6 +221,55 @@ else if (Objects.equals(
...
+
+				Map<String, Object> properties = ddmFormField.getProperties();
+
+				DDMFormFieldOptions rowsDDMFormFieldOptions =
+					(DDMFormFieldOptions)properties.get("rows");
+				DDMFormFieldOptions columnsDDMFormFieldOptions =
+					(DDMFormFieldOptions)properties.get("columns");
+
+				JSONObject jsonObject = JSONFactoryUtil.createJSONObject(
+					valueString);
```
Comment (lfbesada) [inline modules/apps/headless/headless-delivery/headless-delivery-api/src/main/java/com/liferay/headless/delivery/dto/v1_0/util/ContentFieldUtil.java:236]: This should be declared as used, so it should be after selectedValuesJSONObject

**Comment by lfbesada** [comment]
ci:stop

**Review by lfbesada — APPROVED** [review]
LGTM 🎉

**Comment by YurenaC** [comment]
ci:forward

**Comment by Tim-Cao** [comment]
See https://github.com/brianchandotcom/liferay-portal/pull/138151

**Comment by Tim-Cao** [comment]
ci:close

**Pull request closed without merging**